### PR TITLE
Added covariant support to SDL object definitions

### PR DESCRIPTION
**Pull request by bbakerman**
see #962

### CODE REVIEW

File: src/main/java/graphql/schema/idl/SchemaTypeChecker.java [hunk 1]
```diff
@@ -461,6 +461,52 @@ private void checkInterfacesAreImplemented(List<GraphQLError> errors, TypeDefini
         };
     }
 
+    @SuppressWarnings("SimplifiableIfStatement")
+    private boolean isTypeSubTypeOf(TypeDefinitionRegistry registry, Type maybeSubType, Type superType) {
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/SchemaTypeChecker.java]: this method is complex enough to justify maybe a test just for that.

File: src/main/java/graphql/schema/idl/SchemaTypeChecker.java [hunk 2]
```diff
@@ -461,6 +461,52 @@ private void checkInterfacesAreImplemented(List<GraphQLError> errors, TypeDefini
...
+            }
+            return false;
+        }
+        if (maybeSubTypeInfo.isList()) {
+            // If superType is not a list, maybeSubType must also be not a list.
+            return false;
+        }
+
+        // If superType type is an abstract type, maybeSubType type may be a currently
+        // possible object type.
```
Comment (andimarek) [inline src/main/java/graphql/schema/idl/SchemaTypeChecker.java]: maybe this a spec level problem, but I don't see they point to allow union types here: you can't implement a union type

Comment (bbakerman) [reply src/main/java/graphql/schema/idl/SchemaTypeChecker.java]: This come almost word for word from graphql-js

File: src/main/java/graphql/schema/idl/SchemaTypeChecker.java (hunk 1 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/idl/SchemaTypeChecker.java]: done

File: src/main/java/graphql/schema/idl/TypeDefinitionRegistry.java [hunk 3]
```diff
@@ -261,4 +266,120 @@ public boolean hasType(TypeName typeName) {
...
     }
+
+    /**
+     * Returns true if the specified type exists in the registry and is an abstract (Interface or Union) type
+     *
+     * @param type the type to check
+     *
+     * @return true if its abstract
+     */
+    public boolean isAbstractType(Type type) {
```
[the remark this answers is not in the document]
Comment (bbakerman) [reply src/main/java/graphql/schema/idl/TypeDefinitionRegistry.java]: done

**Pull request merged**